packages/medusa/src/services/product-variant-inventory.ts

```diff
@@ -342,20 +375,133 @@ class ProductVariantInventoryService extends TransactionBaseService {
-   * Remove reservation of variant quantity
+   * Adjusts the quantity of reservations for a line item by a given amount.
+   * @param {string} lineItemId - The ID of the line item
+   * @param {string} variantId - The ID of the variant
+   * @param {string} locationId - The ID of the location to prefer adjusting quantities at
+   * @param {number} quantity - The amount to adjust the quantity by
+   */
+  async adjustReservationsQuantityByLineItem(
+    lineItemId: string,
+    variantId: string,
+    locationId: string,
+    quantity: number
+  ): Promise<void> {
+    const manager = this.transactionManager_ || this.manager_
+
+    if (!this.inventoryService_) {
+      const variantServiceTx =
+        this.productVariantService_.withTransaction(manager)
+      const variant = await variantServiceTx.retrieve(variantId, {
+        select: ["id", "inventory_quantity", "manage_inventory"],
+      })
+
+      if (!variant.manage_inventory) {
+        return
+      }
+
+      await variantServiceTx.update(variantId, {
+        inventory_quantity: variant.inventory_quantity - quantity,
+      })
+
+      return
+    }
+    const [reservations, reservationCount] =
+      await this.inventoryService_.listReservationItems(
+        {
+          line_item_id: lineItemId,
+        },
+        {
+          order: { created_at: "DESC" },
+        }
+      )
+
+    if (reservationCount) {
+      let reservation = reservations[0]
+
+      reservation =
+        reservations.find(
+          (r) => r.location_id === locationId && r.quantity >= quantity
+        ) ?? reservation
+
+      const productVariantInventory = await this.retrieve(
+        reservation.inventory_item_id,
+        variantId
+      )
+
+      const reservationQtyUpdate =
+        reservation.quantity - quantity * productVariantInventory.quantity
+
+      if (reservationQtyUpdate === 0) {
+        await this.inventoryService_.deleteReservationItem(reservation.id)
+      } else {
+        await this.inventoryService_.updateReservationItem(reservation.id, {
+          quantity: reservationQtyUpdate,
+        })
+      }
+    }
+  }
+
+  /**
+   * Validate stock at a location for fulfillment items
+   * @param items Fulfillment Line items to validate quantities for
+   * @param locationId Location to validate stock at
+   * @returns nothing if successful, throws error if not
+   */
+  async validateInventoryAtLocation(items: LineItem[], locationId: string) {
+    if (!this.inventoryService_) {
+      return
+    }
+
+    const itemsToValidate = items.filter((item) => item.variant_id)
+
+    for (const item of itemsToValidate) {
+      const pvInventoryItems = await this.listByVariant(item.variant_id!)
+
+      const [inventoryLevels] =
+        await this.inventoryService_.listInventoryLevels({
+          inventory_item_id: pvInventoryItems.map((i) => i.inventory_item_id),
+          location_id: locationId,
+        })
+
+      const pviMap: Record<string, ProductVariantInventoryItem> =
```

Comment:
suggestion(non blocking): a real Map seems easier to read and achieve the same purpose

```ts
const pviMap: Map<string, ProductVariantInventoryItem> = new Map(
        pvInventoryItems.map((pvi) => [pvi.inventory_item_id, pvi])
      )

```